File: src/Functional/Haskell/Proxy.hs
import Data.IORef
import qualified Data.Map.Strict as Map

data Proxy = Proxy
  { backendCreated :: IORef Bool
  , fetchCount :: IORef Int
  , cache :: IORef (Map.Map Int String)
  }

newProxy :: IO Proxy
newProxy = Proxy <$> newIORef False <*> newIORef 0 <*> newIORef Map.empty

remoteGet :: Proxy -> Int -> IO String
remoteGet proxy ident = do
  writeIORef (backendCreated proxy) True
  modifyIORef' (fetchCount proxy) (+ 1)
  pure ("doc(" ++ show ident ++ ")")

proxyGet :: Proxy -> Int -> IO String
proxyGet proxy ident = do
  current <- readIORef (cache proxy)
  case Map.lookup ident current of
    Just value -> pure value
    Nothing -> do
      value <- remoteGet proxy ident
      modifyIORef' (cache proxy) (Map.insert ident value)
      pure value

main :: IO ()
main = do
  proxy <- newProxy
  first <- proxyGet proxy 42
  second <- proxyGet proxy 42
  backend <- readIORef (backendCreated proxy)
  fetches <- readIORef (fetchCount proxy)
  putStrLn $ "backend=" ++ (if backend then "1" else "0") ++ ";fetches=" ++ show fetches ++ ";first=" ++ first ++ ";second=" ++ second

File: src/Functional/Elixir/proxy.exs
defmodule DocumentProxy do
  def start do
    Agent.start_link(fn -> %{created: 0, fetches: 0, cache: %{}} end)
  end

  def get(agent, id) do
    Agent.get_and_update(agent, fn state ->
      case Map.fetch(state.cache, id) do
        {:ok, value} ->
          {value, state}

        :error ->
          created = if state.created == 0, do: 1, else: state.created
          value = "doc(#{id})"
          next = %{
            state
            | created: created,
              fetches: state.fetches + 1,
              cache: Map.put(state.cache, id, value)
          }

          {value, next}
      end
    end)
  end

  def stats(agent), do: Agent.get(agent, &Map.take(&1, [:created, :fetches]))
end

{:ok, proxy} = DocumentProxy.start()
first = DocumentProxy.get(proxy, 42)
second = DocumentProxy.get(proxy, 42)
%{created: created, fetches: fetches} = DocumentProxy.stats(proxy)
IO.puts("backend=#{created};fetches=#{fetches};first=#{first};second=#{second}")
